Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp

```diff
@@ -614,59 +670,67 @@ void GenerateAndAllocateRegisters::generate(CCallHelpers& jit)
                 }
             });
 
-            RegisterSet clobberedRegisters;
-            RegisterSet earlyNextClobberedRegisters;
-            {
-                Inst* nextInst = block->get(instIndex + 1);
-                if (inst.kind.opcode == Patch || (nextInst && nextInst->kind.opcode == Patch)) {
-                    if (inst.kind.opcode == Patch)
-                        clobberedRegisters.merge(inst.extraClobberedRegs());
-                    if (nextInst && nextInst->kind.opcode == Patch)
-                        earlyNextClobberedRegisters.merge(nextInst->extraEarlyClobberedRegs());
-
-                    clobberedRegisters.filter(m_allowedRegisters);
-                    clobberedRegisters.exclude(m_namedDefdRegs);
-                    earlyNextClobberedRegisters.filter(m_allowedRegisters);
-
-                    m_namedDefdRegs.merge(clobberedRegisters);
-                }
+            if (inst.kind.opcode == Patch) {
+                m_earlyClobber.merge(inst.extraEarlyClobberedRegs());
+                m_lateClobber.merge(inst.extraClobberedRegs());
+
+                m_earlyClobber.filter(m_allowedRegisters);
+                m_lateClobber.filter(m_allowedRegisters);
+
+                m_clobberedToClear.merge(m_earlyClobber);
+                m_clobberedToClear.merge(m_lateClobber);
+                m_clobberedToClear.exclude(m_namedDefdRegs);
             }
 
-            auto allocNamed = [&] (const RegisterSet& named, bool isDef) {
+            auto allocNamed = [&] (const RegisterSet& named, Arg::Role role) {
                 for (Reg reg : named) {
                     if (Tmp occupyingTmp = currentAllocation[reg]) {
                         if (occupyingTmp == Tmp(reg))
                             continue;
                     }
 
-                    freeDeadTmpsIfNeeded(); // We don't want to spill a dead tmp.
-                    alloc(Tmp(reg), reg, isDef);
```

Comment:
Should we have an assertion that we are not conflicting with clobbers? (Use + early clobber check and Def + late clobber check).

Reply:
This assertion would fail, since we do things like `clobber(volatileRegisters); patch->resultConstraint = x0`.

So the semantics the register allocators implement is that we allow this to happen, so we don't consider it "clobbered"

Reply:
Oh, I see. It makes sense. So, for named use/def, we have a bit weird semantics with late clobbering, but yeah, for now, it looks fine :)